### PR TITLE
Project Jamie - Paperwork Edition

### PR DESCRIPTION
**Pull request by JamieD1**
Carbon paper:
    Paper bins now have a small menu for taking either normal or carbon paper. Carbon paper is duplicate paper, and once filled out it can be separated into a master and a pink copy paper; good for forms and official paperwork where copies are useful.

Paper Bundle:
    Use one piece of paper with another to clip them together. This generates a paper bundle, which you can browse through by using the relevant added UI buttons. You can also attach photos to these bundles, if desired.

Paper Shredder:
    Paper or ID Cards can now be shredded, creating waste paper. The shredder will fill up with use, and will need to be emptied either with a trash bag, or with an AltClick (Warning: This can make a huge mess)

Fax Machine:
    Every Command role now has a Fax Machine they can use. The fax machine can be used to send anything paper related, including paper bundles (as mentioned above) and photos, to other fax machines. There are also options to send forms or requests to Central Command, if needed. (Yes you can send a picture of your ass to Central - expect __consequences__ if you do, though). Admins are also able to send faxes to station machines!

Bonus - Department Tape:
    The Security and Engineering departments now have police and engineering tape, respectively. These operate somewhat similarly to holobarriers, and create a harmless barricade which can be lifted temporarily with help intent to pass through. It can also be broken with harm intent, or a sharp object. Tapes can also be placed on doors to prevent people from going through them.

:cl:
rscadd: Carbon paper - Get from paperbins, usually for duplicate paperwork
rscadd: Paper Bundles - Put paper together including photos.
rscadd: Paper Shredders - Shreds paper and ID cards
rscadd: Fax Machines - Send Inter departmental and Centcom Faxes
rscadd: Engineering and Security Tape - Usually tape to help cordon off areas
/:cl:

### CODE REVIEW

File: code/modules/paperwork/faxmachine.dm [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	rcvdcopy.loc = null //hopefully this shouldn't cause trouble
+	adminfaxes += rcvdcopy
+	
+	//message badmins that a fax has arrived
+	switch(destination)
+		if ("Central Command")
+			message_admins(sender, "CENTCOMM FAX", rcvdcopy, "CentcommFaxReply", "#006100")
+		if ("Sol Government")
+			message_admins(sender, "SOL GOVERNMENT FAX", rcvdcopy, "CentcommFaxReply", "#1F66A0")
+			//message_admins(sender, "SOL GOVERNMENT FAX", rcvdcopy, "SolGovFaxReply", "#1F66A0")
```
Comment (JamieD1) [inline code/modules/paperwork/faxmachine.dm]:
```suggestion
			message_admins(sender, "CENTCOM FAX", rcvdcopy, "CentcomFaxReply", "#006100")
		if ("Sol Government")
			message_admins(sender, "SOL GOVERNMENT FAX", rcvdcopy, "CentcomFaxReply", "#1F66A0")
			//message_admins(sender, "SOL GOVERNMENT FAX", rcvdcopy, "SolGovFaxReply", "#1F66A0")
```

**Comment by Bibby0110** [comment]
Are ya still working on this?

**Comment by maxion12345** [comment]
fax machine when

**Comment by Bibby0110** [comment]
Stale.

**Comment by maxion12345** [comment]
Where is my fax machine

**Comment by Tipy1802** [comment]
Good luck o7

**Comment by Bibby0110** [comment]
Stale for more than a month.
Reopen when you continue working on it :)

File: code/game/objects/items/barriertape.dm [hunk 2]
```diff
@@ -0,0 +1,183 @@
+//Define all tape types in policetape.dm
```
Comment (ToasterBiome) [inline code/game/objects/items/barriertape.dm]: sir this is barriertape.dm

File: code/game/objects/items/barriertape.dm [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+/obj/item/barrier_taperoll
+	name = "barrier tape roll"
+	icon = 'icons/obj/barriertape.dmi'
+	icon_state = "rollstart"
+	w_class = 2.0
```
Comment (ToasterBiome) [inline code/game/objects/items/barriertape.dm]:
```suggestion
	w_class = WEIGHT_CLASS_SMALL
```

File: code/game/objects/items/barriertape.dm [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			return 1
+		if(allowed(mover))
+			return 1
+
+/obj/structure/barrier_tape/Bumped(atom/movable/AM)
+	. = ..()
+	if(iscarbon(AM))
+		add_fingerprint(AM)
+		if(lifted == FALSE)
+			to_chat(AM, "<span class='warning'>You are not supposed to go past [src]... (You can break the tape with something sharp or lift the tape with HELP intent)</span>")
```
Comment (ToasterBiome) [inline code/game/objects/items/barriertape.dm]:
```suggestion
			to_chat(AM, "<span class='warning'>You are not supposed to go past [src]...</span>")
```

I think the explanation would be better on a close examination or something, not just hitting it.

File: code/modules/paperwork/faxmachine.dm [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+GLOBAL_LIST_INIT(admin_departments, list("Central Command"))
+GLOBAL_LIST_EMPTY(alldepartments)
+GLOBAL_LIST_EMPTY(adminfaxes)
+
+/obj/machinery/photocopier/faxmachine
+	name = "fax machine"
+	icon = 'icons/obj/library.dmi'
+	icon_state = "fax"
+	insert_anim = "faxsend"
+	req_one_access = list(ACCESS_LAWYER, ACCESS_HEADS, ACCESS_ARMORY, ACCESS_FORENSICS_LOCKERS) //Warden needs to be able to Fax solgov too.
```
Comment (ToasterBiome) [inline code/modules/paperwork/faxmachine.dm]: what is solgov?

File: code/modules/paperwork/paper_bundle.dm [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	pressure_resistance = 1
+	attack_verb = list("bapped")
+	var/amount = 0 //Amount of items clipped to the paper
+	var/page = 1
+	var/screen = 0
+	/// If this was sent via admin fax, allows anyone to see/interact with it
+	var/admin_faxed = FALSE
+
+
+/obj/item/paper_bundle/attackby(obj/item/W as obj, mob/user as mob)
```
Comment (ToasterBiome) [inline code/modules/paperwork/paper_bundle.dm]: do we use this typing elsewhere in our code? Only other place I see it is modular computer

File: code/modules/paperwork/papershredder.dm [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+			paperamount += 1
+			update_icon()
+			return
+	
+	if(shred_amount && try_insert(user, shred_amount))
+		qdel(W)
+
+
+
+/obj/machinery/papershredder/verb/emtpy()
```
Comment (ToasterBiome) [inline code/modules/paperwork/papershredder.dm]: emtpy

File: code/modules/paperwork/papershredder.dm [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		icon_state = "papershredder2"
+	if(paperamount == 5||paperamount == 6)
+		icon_state = "papershredder3"
+	if(paperamount == 7||paperamount == 8)
+		icon_state = "papershredder4"
+	if(paperamount == 9||paperamount == 10)
+		icon_state = "papershredder5"
+	return
+
+/obj/item/shreddedp
```
Comment (ToasterBiome) [inline code/modules/paperwork/papershredder.dm]:
```suggestion
/obj/item/shreddedpaper
```

shreddedp is a terrible name

File: code/modules/paperwork/papershredder.dm [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+			if(user.get_inactive_hand_index() == src)
+				user.dropItemToGround(src)
+
+			new /obj/effect/decal/cleanable/ash(src.loc)
+			qdel(src)
+
+		else
+			to_chat(user, "<span class='warning'>You must hold \the [P] steady to burn \the [src].</span>")
+
+/obj/item/shreddedp/proc/FireBurn()
```
Comment (ToasterBiome) [inline code/modules/paperwork/papershredder.dm]: We have /obj/item/burn() in code/game/objects/items.dm

**Review by ToasterBiome — DISMISSED** [review]

File: code/game/objects/items/barriertape.dm (hunk 2 above)
Comment (JamieD1) [reply code/game/objects/items/barriertape.dm]: lies

File: code/game/objects/items/barriertape.dm (hunk 4 above)
Comment (JamieD1) [reply code/game/objects/items/barriertape.dm]: The idea was for the first month or so put it in for dumb people

File: code/modules/paperwork/faxmachine.dm (hunk 5 above)
Comment (JamieD1) [reply code/modules/paperwork/faxmachine.dm]: oops

File: code/modules/paperwork/paper_bundle.dm (hunk 6 above)
Comment (JamieD1) [reply code/modules/paperwork/paper_bundle.dm]: oops

File: code/modules/paperwork/papershredder.dm (hunk 7 above)
Comment (JamieD1) [reply code/modules/paperwork/papershredder.dm]: emtpy

File: code/modules/paperwork/papershredder.dm (hunk 8 above)
Comment (JamieD1) [reply code/modules/paperwork/papershredder.dm]: baimou is a terrible name

File: code/modules/admin/topic.dm [hunk 10]
```diff
@@ -2264,6 +2264,58 @@
...
+						// Stamps
+						var/image/stampoverlay = image('icons/obj/bureaucracy.dmi')
+						stampoverlay.icon_state = "paper_stamp-cent"
+						if(!P.stamped)
+							P.stamped = new
+						P.stamped += /obj/item/stamp
+						P.overlays += stampoverlay
+						P.stamps += "<HR><i>This paper has been stamped by the Central Command Quantum Relay.</i>"
+
+				to_chat(owner, "Message reply to transmitted successfully.")
```
Comment (Anvilman6) [inline code/modules/admin/topic.dm]: Message reply to what?

File: code/game/objects/items/barriertape.dm [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+/obj/structure/barrier_tape/attackby(obj/item/W as obj, mob/user as mob)
+	breaktape(W, user)
+
+/obj/structure/barrier_tape/attack_hand(mob/user as mob)
+	if (user.a_intent == "help" )
+		user.visible_message("<span class='notice'>[user] lifts [src], allowing passage.</span>")
```
Comment (Anvilman6) [inline code/game/objects/items/barriertape.dm]: [user] lifts the [src]

File: code/modules/paperwork/paperbin.dm [hunk 12]
```diff
@@ -74,6 +75,10 @@
 		bin_pen = null
 		update_icon()
 	else if(total_paper >= 1)
+		response = alert(user, "Do you take regular paper, or Carbon copy paper?", "Paper type request", "Regular", "Carbon-Copy", "Cancel")
```
Comment (Anvilman6) [inline code/modules/paperwork/paperbin.dm]: Keep spelling the same here: "carbon copy paper". Can the hyphenated version in the options be made not hypenated?

File: code/modules/paperwork/paperbin.dm [hunk 13]
```diff
@@ -74,6 +75,10 @@
...
+		if (response != "Regular" && response != "Carbon-Copy")
```
Comment (Anvilman6) [inline code/modules/paperwork/paperbin.dm]: If hyphen can be removed, strike it here too

File: code/modules/paperwork/paperbin.dm [hunk 14]
```diff
@@ -82,11 +87,14 @@
...
-				if(prob(30))
-					P.info = "<font face=\"[CRAYON_FONT]\" color=\"red\"><b>HONK HONK HONK HONK HONK HONK HONK<br>HOOOOOOOOOOOOOOOOOOOOOONK<br>APRIL FOOLS</b></font>"
-					P.rigged = 1
+			if(response == "Regular")
+				P = new papertype(src)
+				if(SSevents.holidays && SSevents.holidays[APRIL_FOOLS])
+					if(prob(30))
+						P.info = "<font face=\"[CRAYON_FONT]\" color=\"red\"><b>HONK HONK HONK HONK HONK HONK HONK<br>HOOOOOOOOOOOOOOOOOOOOOONK<br>APRIL FOOLS</b></font>"
+						P.rigged = 1
+			else if (response == "Carbon-Copy")
```
Comment (Anvilman6) [inline code/modules/paperwork/paperbin.dm]: Above hyphen note

File: code/modules/paperwork/papershredder.dm [hunk 15]
```diff
@@ -0,0 +1,173 @@
+/obj/machinery/papershredder
+	name = "Paper Shredder"
+	desc = "For those documents you don't want seen."
```
Comment (Anvilman6) [inline code/modules/paperwork/papershredder.dm]: "Disposes of papers you don't want seen, as well as IDs that are no longer needed."

**Review by Anvilman6 — CHANGES_REQUESTED** [review]
Semantics

File: code/game/objects/items/storage/belt.dm [hunk 16]
```diff
@@ -42,9 +42,9 @@
 /obj/item/storage/belt/utility/ComponentInitialize()
 	. = ..()
 	var/datum/component/storage/STR = GetComponent(/datum/component/storage)
-	STR.max_items = 10
```
Comment (ToasterBiome) [inline code/game/objects/items/storage/belt.dm:45]: atomize this into it's own PR

**Review by ToasterBiome — APPROVED** [review]
dont like the unseen belt buffs, but code looks good enough

File: code/game/objects/items/barriertape.dm [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+			if(cur.density || istype(cur, /turf/open/space))
+				can_place = FALSE
+			else
+				for(var/obj/O in cur)
+					if(!istype(O, /obj/structure/barrier_tape) && O.density)
+						can_place = FALSE
+						break
+			cur = get_step_towards(cur,end)
+		if (!can_place)
+			to_chat(usr, "<span class='warning'>You can't run \the [src] through that!</notice>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+		start = get_turf(src)
+		to_chat(usr, "<span class='notice'>You place the first end of the [src].</span>")
+		icon_state = "[icon_base]_stop"
+		placing = TRUE
+	else
+		placing = FALSE
+		icon_state = "[icon_base]_start"
+		end = get_turf(src)
+		if(start.y != end.y && start.x != end.x || start.z != end.z)
+			to_chat(usr, "<span class='notice'>[src] can only be laid horizontally or vertically.</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+			for(var/obj/structure/barrier_tape/Ptest in cur)
+				if(Ptest.tape_dir == dir)
+					tapetest = TRUE
+			if(!tapetest)
+				var/obj/structure/barrier_tape/P = new tape_type(cur)
+				P.icon_state = "[P.icon_base]_[dir]"
+				P.tape_dir = dir
+			cur = get_step_towards(cur,end)
+	//is_blocked_turf(var/turf/T)
+		to_chat(usr, "<span class='notice'>You finish placing the [src].</span>")	//Git Test
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+/obj/item/barrier_taperoll/afterattack(var/atom/A, var/mob/user, proximity)
+	if (proximity && istype(A, /obj/machinery/door/airlock))
+		var/turf/T = get_turf(A)
+		var/obj/structure/barrier_tape/P = new tape_type(T.x,T.y,T.z)
+		P.loc = locate(T.x,T.y,T.z)
+		P.icon_state = "[icon_base]_door"
+		P.layer = 3.2
+		to_chat(user, "<span class='notice'>You finish placing the [src].</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+			return 1
+		if(allowed(mover))
+			return 1
+
+/obj/structure/barrier_tape/Bumped(atom/movable/AM)
+	. = ..()
+	if(iscarbon(AM))
+		add_fingerprint(AM)
+		if(lifted == FALSE)
+			to_chat(AM, "<span class='warning'>You are not supposed to go past [src]... (You can break the tape with something sharp or lift the tape with HELP intent)</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+
+/obj/structure/barrier_tape/attackby(var/obj/item/W, var/mob/user)
+	breaktape(W, user)
+
+/obj/structure/barrier_tape/attack_hand(var/mob/user)
+	if (user.a_intent == "help" )
+		user.visible_message("<span class='notice'>[user] lifts [src], allowing passage.</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+		crumple()
+		lifted = TRUE
+		addtimer(VARSET_CALLBACK(src, lifted, FALSE), 20)
+	else
+		breaktape(null, user)
+
+/obj/structure/barrier_tape/proc/breaktape(var/obj/item/W, var/mob/user)
+	if(user.a_intent == INTENT_HELP && W && !W.is_sharp() && allowed(user))
+		to_chat(user, "<span class='warning'>You can't break the [src] with that!</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/game/objects/items/barriertape.dm [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+		return
+	user.visible_message("<span class='notice'>[user] breaks the [src]!</span>")
```
Comment (adamsong) [inline code/game/objects/items/barriertape.dm]: span macro

File: code/modules/admin/topic.dm [hunk 25]
```diff
@@ -2264,6 +2264,58 @@
 	else if(href_list["beakerpanel"])
 		beaker_panel_act(href_list)
 
+	else if(href_list["AdminFaxView"])
+		var/obj/info = locate(href_list["AdminFaxView"])
```
Comment (adamsong) [inline code/modules/admin/topic.dm]:
```suggestion
		var/obj/info = locate(href_list["AdminFaxView"]) in GLOB.adminfaxes
```

File: code/modules/admin/topic.dm [hunk 26]
```diff
@@ -2264,6 +2264,58 @@
...
+		if(info)
+			info.examine(usr, TRUE)
+
+	else if(href_list["CentcomFaxReply"])
+		var/mob/living/carbon/human/H = locate(href_list["CentcomFaxReply"])
+		var/obj/machinery/photocopier/faxmachine/fax = locate(href_list["originfax"])
```
Comment (adamsong) [inline code/modules/admin/topic.dm]:
```suggestion
		var/obj/machinery/photocopier/faxmachine/fax = locate(href_list["originfax"]) in GLOB.allfaxes
```

File: code/modules/admin/topic.dm [hunk 27]
```diff
@@ -2264,6 +2264,58 @@
 	else if(href_list["beakerpanel"])
 		beaker_panel_act(href_list)
 
+	else if(href_list["AdminFaxView"])
+		var/obj/info = locate(href_list["AdminFaxView"])
+		if(info)
+			info.examine(usr, TRUE)
+
+	else if(href_list["CentcomFaxReply"])
+		var/mob/living/carbon/human/H = locate(href_list["CentcomFaxReply"])
```
Comment (adamsong) [inline code/modules/admin/topic.dm:2273]:
```suggestion
		var/mob/living/carbon/human/H = locate(href_list["CentcomFaxReply"])
		if(!istype(H)) return
```

File: code/modules/paperwork/carbonpaper.dm [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+		copy.copy_overlays(src, TRUE)
+
+		copied = TRUE
+		copy.copied = TRUE
+		copy.iscopy = TRUE
+		update_icon()
+		copy.update_icon()
+		usr.put_in_hands(copy)
+	else
+		to_chat(usr, "<span class='warning'>There are no more carbon copies attached to this paper!</span>")
```
Comment (adamsong) [inline code/modules/paperwork/carbonpaper.dm]: span macro

File: code/modules/paperwork/faxmachine.dm [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+	var/auth_name
+	var/sendcooldown = 0 // to avoid spamming fax messages
+	var/department = "Unknown" // our department
+	var/destination = "Central Command" // the department we're sending to
+
+/obj/machinery/photocopier/faxmachine/Initialize()
+	. = ..()
+	GLOB.allfaxes += src
+
+	if( !(("[department]" in GLOB.alldepartments) || ("[department]" in GLOB.admin_departments)) )
```
Comment (adamsong) [inline code/modules/paperwork/faxmachine.dm]:
```suggestion
	if( !((department in GLOB.alldepartments) || (department in GLOB.admin_departments)) )
```

File: code/modules/paperwork/faxmachine.dm [hunk 30]
```diff
@@ -0,0 +1,188 @@
...
+/obj/machinery/photocopier/faxmachine/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+	. = TRUE
+	switch(action)
+		if("send")
+			if(!copier_empty())
+				if(sendcooldown - world.time > 0)
+					to_chat(usr, "<span class='warning'>Transmitter recharging</span>")
```
Comment (adamsong) [inline code/modules/paperwork/faxmachine.dm:56]: span macro

File: code/modules/paperwork/faxmachine.dm [hunk 31]
```diff
@@ -0,0 +1,188 @@
...
+				if (destination in GLOB.admin_departments)
+					INVOKE_ASYNC(src, .proc/send_admin_fax, usr, destination)
+				else
+					INVOKE_ASYNC(src, .proc/sendfax, destination)
+			return
+		if("remove")
+			if(copy)
+				copy.loc = usr.loc
+				usr.put_in_hands(copy)
+				to_chat(usr, "<span class='notice'>You take \the [copy] out of \the [src].</span>")
```
Comment (adamsong) [inline code/modules/paperwork/faxmachine.dm:67]: span macro

File: code/modules/paperwork/faxmachine.dm [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+				copy = null
+				return
+			if(photocopy)
+				photocopy.loc = usr.loc
+				usr.put_in_hands(photocopy)
+				to_chat(usr, "<span class='notice'>You take \the [photocopy] out of \the [src].</span>")
```
Comment (adamsong) [inline code/modules/paperwork/faxmachine.dm:73]: span macro

File: code/modules/paperwork/papershredder.dm [hunk 33]
```diff
@@ -0,0 +1,166 @@
...
+				var/ran = rand(1,3)
+				if(ran == 1)
+					SP.color = "#BABABA"
+				if(ran == 2)
+					SP.color = "#7F7F7F"
+				if(ran == 3)
+					SP.color = null
+				paperamount -=1
+			update_icon()
+			to_chat(user, "<span class='warning'>The [src] was too full and shredded paper goes everywhere!</span>")
```
Comment (adamsong) [inline code/modules/paperwork/papershredder.dm]: span macro

File: code/modules/paperwork/papershredder.dm [hunk 34]
```diff
@@ -0,0 +1,166 @@
...
+		else
+			to_chat(user, "<span class='warning'>The [src] is full please empty it before you continue.</span>")
```
Comment (adamsong) [inline code/modules/paperwork/papershredder.dm]: span macro

File: code/modules/paperwork/papershredder.dm [hunk 35]
```diff
@@ -0,0 +1,166 @@
...
+					SP.color = "#BABABA"
+				if(ran == 2)
+					SP.color = "#7F7F7F"
+				if(ran == 3)
+					SP.color = null
+				STR.handle_item_insertion(SP)
+				paperamount -=1
+				update_icon()
+			else
+				to_chat(user, "<span class='warning'>The [W] is full.</span>")
```
Comment (adamsong) [inline code/modules/paperwork/papershredder.dm]: span macro

File: code/modules/paperwork/papershredder.dm [hunk 36]
```diff
@@ -0,0 +1,166 @@
...
+	if (istype(W, /obj/item/paper))
+		shred_amount = 1
+	else if(istype(W, /obj/item/photo))
+		shred_amount = 1
+	else if(istype(W, /obj/item/newspaper))
+		shred_amount = 3
+	else if(istype(W, /obj/item/card/id))
+		shred_amount = 3
+	else if(istype(W, /obj/item/paper_bundle))
+		shred_amount = 3
```
Comment (adamsong) [inline code/modules/paperwork/papershredder.dm:52]: Should this be based on amount?

File: code/modules/paperwork/papershredder.dm [hunk 37]
```diff
@@ -0,0 +1,166 @@
...
+					SP.color = null
+				STR.handle_item_insertion(SP)
+				paperamount -=1
+				update_icon()
+			else
+				to_chat(user, "<span class='warning'>The [W] is full.</span>")
+				return
+	else if(istype(W, /obj/item/shreddedpaper))
+		if(paperamount == max_paper)
+			to_chat(user, "<span class='warning'>The [src] is full please empty it before you continue.</span>")
```
Comment (adamsong) [inline code/modules/paperwork/papershredder.dm]: span macro

File: code/modules/paperwork/photocopier.dm [hunk 38]
```diff
@@ -347,3 +315,69 @@
...
+			toner--
+		return c
+
+
+/obj/machinery/photocopier/proc/photocopy(var/obj/item/photo/photocopy)
+	. = new /obj/item/photo (loc, photocopy.picture.Copy(greytoggle == "Greyscale"? TRUE : FALSE))
+	toner -= 5	//photos use a lot of ink!
+	if(toner < 0)
+		toner = 0
+		visible_message("<span class='notice'>A red light on \the [src] flashes, indicating that it is out of toner.</span>")
```
Comment (adamsong) [inline code/modules/paperwork/photocopier.dm:359]: span macro

File: code/modules/paperwork/photocopier.dm [hunk 39]
```diff
@@ -347,3 +315,69 @@
...
+
+//If need_toner is 0, the copies will still be lightened when low on toner, however it will not be prevented from printing. TODO: Implement print queues for fax machines and get rid of need_toner
+/obj/machinery/photocopier/proc/bundlecopy(var/obj/item/paper_bundle/bundle, var/need_toner=1)
+	var/obj/item/paper_bundle/p = new /obj/item/paper_bundle (src)
+	for(var/obj/item/W in bundle)
+		if(toner <= 0 && need_toner)
+			toner = 0
+			visible_message("<span class='notice'>A red light on \the [src] flashes, indicating that it is out of toner.</span>")
```
Comment (adamsong) [inline code/modules/paperwork/photocopier.dm:367]: span macro

**Review by adamsong — CHANGES_REQUESTED** [review]

**Review by adamsong — APPROVED** [review]

**Review by nesisknesis — APPROVED** [review]

**Review by Altoids1 — CHANGES_REQUESTED** [review]
You're missing a colon in your changelog.

**Comment by stwordcitizen** [comment]
Engi tape is way too broken, any one can just spam down a huge line of 2 tile long tape lines over and over to escape being chased by someone, also the fact that tape is infinite doesn't help either.

**Pull request merged**